### ISSUE TITLE
Unable to understand how to use the code for CBL

### ISSUE THREAD

**Issue by it-is-lokesh** [dywu98/CBL-Conditional-Boundary-Loss#4]
Hi, thank you for such a great methodology. We are trying to understand your code given in the repository but unable to comprehend what the loss function takes in as parameters. Can you kindly give the path to the loss function code along with the description/steps to use the same in a custom segmentation model? 
Thanks

**Comment by dywu98** [comment]
Thanks for the attention on our work. 

First, we recommend you to directly use the [maskformer](https://github.com/dywu98/CBL-Conditional-Boundary-Loss/tree/main/maskformer) or [mask2former](https://github.com/dywu98/CBL-Conditional-Boundary-Loss/tree/main/mask2former) code, since we have provide detailed instruction on how to train a maskformer or mask2former model with our CBL loss. Be sure to run the `train_net_biou.py` with CBL instead of the original `train_net.py` for the plain maskformer and mask2former.

Second, the loss function takes the predicted logits, mid feature maps, ground truth, and boundary mask as inputs. 
Each element in the boundary mask is 0 or 1 which indicates whether the pixel is or is not located in the boundary areas.
You can find the CBL code for mask2former at [here](https://github.com/dywu98/CBL-Conditional-Boundary-Loss/blob/9adccfe7a135647301978074fba4f92654aad9ac/mask2former/mask2former/maskformer_model.py#L246).
We generate the boundary mask for mask2former with the data processing pipeline in this [code](https://github.com/dywu98/CBL-Conditional-Boundary-Loss/blob/9adccfe7a135647301978074fba4f92654aad9ac/mask2former/mask2former/data/dataset_mappers/mask_former_semantic_dataset_mapper_biou.py#L168)
For mmsegmentation, your can refer to [here](https://github.com/dywu98/CBL-Conditional-Boundary-Loss/blob/main/CBLocr_head.py#L213) with [the code to generate the boundary mask](https://github.com/dywu98/CBL-Conditional-Boundary-Loss/blob/9adccfe7a135647301978074fba4f92654aad9ac/biou.py#L83). The dilation rate in it controls the width of the boundary.

The steps to train a OCRNet model in mmsegmentation is listed in the readme.md. Sorry for the poor and confusing description. If you need the full mmsegmentation code, please leave an e-mail address so that I can send you a version for Swin-B-UpperNet+CBL which can run directly but no longer supported by the latest openmmlab segmentation repo.

**Comment by it-is-lokesh** [comment]
Thanks a lot for the reply. It solves all my queries and I request you to add the links to the CBL code for mask2former and boundary mask extractor in the README file, so that I helps anyone in the future who is using the code. 
Please send the code for Swin-B-UpperNet+CBL, it would help us in understanding the loss.
Here is my email id: g.sailokesh9@gmail.com
Thanks again!

**Comment by it-is-lokesh** [comment]
Can you kindly explain what the dictionary given as Batched_input to the MaskFormer class contains? Specifically, what are "instances", "sem_seg" and "sem_seg_boundary"? Also what does the output dictionary consist of? Thank you!

**Comment by it-is-lokesh** [comment]
Can you kinldy explain what is outputs['mask_features'] in criterion.py? I am confused which variable is the logits from backbone and which is the actual mask prediction (after applying softmax)!

**Comment by it-is-lokesh** [comment]
Also, can you please provide link to the A2P&N loss computation, I couldnt find its implementation in [criterion.py](https://github.com/dywu98/CBL-Conditional-Boundary-Loss/blob/main/mask2former/mask2former/modeling/criterion.py). 
Thanks!

**Comment by yamyCD** [comment]
Hello,

Thank you for the detailed response. I am also interested in obtaining the complete code, especially the parts related to mask2former and boundary mask extractor for CBL. It would be very helpful for understanding the loss function and implementing it.

Could you please send me the Swin-B-UpperNet+CBL version of the code as well? My email is: jlove1146@163.com.

Thank you very much!

**Comment by dywu98** [comment]
> Hello,
> 
> Thank you for the detailed response. I am also interested in obtaining the complete code, especially the parts related to mask2former and boundary mask extractor for CBL. It would be very helpful for understanding the loss function and implementing it.
> 
> Could you please send me the Swin-B-UpperNet+CBL version of the code as well? My email is: [jlove1146@163.com](mailto:jlove1146@163.com).
> 
> Thank you very much!

Code sent. Please check your email. Feel free to ask if you have any question.

**Comment by lovelyasfds** [comment]
Hello,

Thank you for the detailed response. I am also interested in obtaining the complete code, especially the parts related to mask2former and boundary mask extractor for CBL. It would be very helpful for understanding the loss function and implementing it.

Could you please send me the Swin-B-UpperNet+CBL version of the code as well? My email is: [1106830647@qq.com](mailto:jlove1146@163.com).

Thank you very much!

**Comment by dywu98** [comment]
> Hello,
> 
> Thank you for the detailed response. I am also interested in obtaining the complete code, especially the parts related to mask2former and boundary mask extractor for CBL. It would be very helpful for understanding the loss function and implementing it.
> 
> Could you please send me the Swin-B-UpperNet+CBL version of the code as well? My email is: [1106830647@qq.com](mailto:jlove1146@163.com).
> 
> Thank you very much!

Thanks for your attention on our work! Please note that we have provided the code of CBL+mask2former in this repository. You can check here [mask2former+CBL](https://github.com/dywu98/CBL-Conditional-Boundary-Loss/tree/main/mask2former) for the implementation, which should be easy to use following the instructions.

We have also send the Swin-B-UpperNet+CBL code via email. Feel free to ask if you have further questions.

**Comment by Lz-Web-git** [comment]
Hello author，Could you please send me the Swin-B-UpperNet+CBL version of the code as well? My email is: 15591896137@163.com. Thanks.

**Comment by dywu98** [comment]
> Hello author，Could you please send me the Swin-B-UpperNet+CBL version of the code as well? My email is: [15591896137@163.com](mailto:15591896137@163.com). Thanks.

Sorry for the late reply. The code of Swin-B-UpperNet+CBL has been sent to your email. Feel free to ask if you encounter reproduction issues.

**Comment by Yayan1128** [comment]
Hello,

Thank you for the detailed response.  Could you please send me the Swin-B-UpperNet+CBL version of the code as well? My email is: wangyanyan_neu@163.com

Thank you very much!
